### DISCUSSION TITLE
Other language

### DISCUSSION THREAD

**Discussion by X-Ryl669** [dscripka/openWakeWord#52] [General]
Well... as the title says. I'd like to know what are the limitation to have OWW working on a different language?

## Training

For training, I understand that it's hard to find multi-speaker models in different language than English. Yet, there are some, so maybe it would be ok to list some languages where it might work:

- [X] English 
- [  ] French
- [  ] German
- [  ] Spanish
- [  ] Chinese

I'll add more language if multispeaker model exist ?

I'm wondering if it's possible to clone a voice from one existing model with the feature of the user's own voice (and any other user's like in Common Voice dataset)? If yes, how to do that?

## Voice activity detection

I'm probably wrong here, but I don't think VAD should be language specifc, does it ?

## Inference

I don't think the language has an influence on the interference run. Please correct if I'm wrong.

**Comment by dscripka** [comment]
Thanks for starting the discussion, @X-Ryl669!

Right now the biggest limitations for multilingual support are:

1) Multi-speaker text-to-speech models in other languages, in particular VITs based models (e.g., [Piper](https://github.com/rhasspy/piper)) as that works the best to train new models. This should be possible to do with Piper, and it something that we are hoping to test soon. Voice cloning and voice conversion, as you mention, is also a possible option that we are hoping to explore.

2) The pre-trained audio embedding model from Google may have only been trained on English data (we actually don't know, the paper doesn't specify). If this is true, the features that in produces may not be as effective on other languages. This is just a theory, though, and it is certainly worth testing once a good multi-speaker multi-lingual VITs model is available.

Outside of this, though, there are not many barriers. VAD works fine across many different languages, and the inference process would be identical once the model is trained.

One potential way that community support would be very beneficial, though, is having native speakers test any non-english models that we train to ensure that the wake word is properly detected with a correct and realistic pronunciation. This is the best way to ensure that the model does indeed perform well on the target language.

**Reply by edurenye** [reply]
We should make the TTS multilingual and phonetically based, you should input a texts using [IPA](https://en.wikipedia.org/wiki/International_Phonetic_Alphabet) and then it should be able to listen to any language.

A unified multilingual mode would work much better than a single language trained model.
[This study provides strong evidence that unified representations should be the preferred paradigm when building multilingual NTTS systems.](https://arxiv.org/pdf/2207.01547.pdf)

**Reply by X-Ryl669** [reply]
I don't understand here. I think I've learnt that you can't speak a language by just concatenating phonemes. A language is more than just phoneme, the fluency, the pitch rate, and many other vocal artifact are required to perform a good representation. I was under the impression that the opposite is true, you can't actually be as accurate with just a Speech To Phoneme model. 

The linked paper isn't really using phoneme here, but phonetic tokens. It's using a lot more phonetic tokens that they are phonemes, and they've shown that the more "token" you use the better the results. This seems to go against the conclusion that it's possible to convert a wakeword to a sequence of phonemes that a universal model could match upon.

Or: it would work only if there was enough phonetic tokens, but in that case, you still have to split your wake word sequence in a token sequence which might take as much work as the ASR algorithm. In that case, you don't gain anything, just run the ASR algorithm and match if it starts with the wake words.

**Reply by edurenye** [reply]
Those where 2 different comments, sorry I was not clear enough.

First I was just wondering if we could use phonetics to input the target word in the trining (I'm not sure exactly how that is used), but maybe is irrelevant. I was wondering that maybe that is why you may need in a multilangual model, but I do not know.

The other comment was about using a multilangual model because acording to that paper that works better than using a different model for each language, or at least that is what I understood.

**Reply by StuartIanNaylor** [reply]
Likely its true about the audio embedding model as it could be English and not contain certain types of language embedding.
I havent looked but presuming your just using a simple Silence (Noise), KW & !KW head classification?
The balance of the phonetics in !KW could well effect accuracy and never managed this with real data due to a lack of it.
Likely though you still get the same problem of a all-in-one !KW as the collation of phonetics as sylables creates quite clear spectra.
When you are trying to add words its always a see-saw that might tip the balance of other words in terms of phonetic balance.
When you bang in an all-in-one !KW with all those vectors its very hard to at least visualise what all those vectors create and often you get false positives not because you have a KW but because the spectra you have doesn't fit !KW.
So the softmax of the classes output flips the otherway even if simularity is low because !kw & Noise are near zero.
Guess you could create a custom softmax that requires an minimum tally for KW...
I read adding classes incurs little extra weight and you might garner better accuracy by splitting !KW into multiple phonetically sylable grouped phonetic !KW classes.
In fact likely you can cheat especially with synthetic as I used to try an squeeze the KW with 'sounds-like' !KW but always struggled with the absence of real data. Just make sure words are not repeated in other classes.
Doing so especially closer sounds-like classes will force the model to train harder and make it less prone to overfitting.
Likely !kw would benefit from some sort of https://en.wikipedia.org/wiki/Sonority_Sequencing_Principle but that level of language skills likely needs a certain amount of expertease of how to partition classes into weighted sonority.
Here with KW we are not talking about language as its basically an image classification model and the sylable groups of phones does have a direct link to spectra.
There is a sonority_hierarchy as all phones are not equal https://github.com/StuartIanNaylor/ProjectEars/blob/main/dataset/ml-commons/create-db-syllable.py as some are just accents to major phones, but you can.
Actually thinking about it I doubt the embedding model is English only as I have a gut feeling they used Voxceleb but if you read.
https://arxiv.org/pdf/2002.01322.pdf
> 3.1. Embedding Model
> The embedding model converts a stream of audio into a
> stream of 96-dimensional feature vectors, one every 80 ms.
> To ensure that the resulting embedding is useful for arbitrary sets of keywords, we took 5000 keywords (most of
> them are actually 2-3 words long) and split them into random groups of 40 keywords. The resulting 125 groups
> of keywords are used to train 125 keyword spotting models with shared weights for the embedding model part (see
> Figure 1). We used roughly 200 million 2-second audio
> clips from YouTube for training, of which 100 million contained the target keywords and the other 100 million were
> used as non-target examples. This embedding model is
> available on Tensorflow Hub for public use. The embedding model is trained using TensorFlow [1] on 20 GPUs
> for 2 days and is available on TensorFlow Hub (https:
> //tfhub.dev/google/speech_embedding/1) for
> reuse.

They have took a subset 5000 keywords so depending on those keywords they could likely be English words and also users might be using KW that there is no corresponding embedding. But doesn't sort of say much about the embedding model though.
At a guess because really its very widely used and they name drop youtube I think its the multilingual https://www.robots.ox.ac.uk/~vgg/data/voxceleb/ 1&2 dataset for the embedding model as "20 GPUs
for 2 days" is pretty extreme and likely translates to the whole 2000 hours+ of Voxceleb 1&2.
Its is 'multilingual' but like Whisper heavily biased to USA https://github.com/cyrta/voxceleb

So its sort of like Whisper as it is supposedly multilingual but the embedding model uses a dataset that is massively biased to US/UK English with some languages getting such sparse data you have to use the term 'Suposedly'
When creating synthetic data through a phonetic based TTS even if already there is strong correlation to spectra by sylable grouped phones of a sonority hierarchy, using a phonetic based TTS very much enforces it.

https://arxiv.org/abs/2210.17016 Wenet have some great stuff on this kind of thing.

**Comment by X-Ryl669** [comment]
Can you give a rough estimate of the corpus required for training a new language? 
Are we speaking of 100 utterances? 1000? more?
How many different speakers?

**Reply by dscripka** [reply]
This depends on many different factors, and there isn't (to my knowledge) a strong consensus on the volume of data required for training. For the VITs TTS models that seem to work well with openWakeWord, I think at minimum ~100 speakers and several hours of speech per speaker is ideal. However, experiments may demonstrate that less is still sufficient.

**Reply by StuartIanNaylor** [reply]
I am having a guestimate but from how popular voxceleb is as a dataset and they name drop Youtube and say it took 2days on 20 GPU's to train the embedding model....
Voxceleb is 7000+ speakers, 1m utterances, 2000+ hours but prob is as a multilingual dataset its completely skewed to US/UK english.
So likely not that big but much more than 100 utterances? 1000?
If you where going to do French then likely https://datashare.ed.ac.uk/handle/10283/2353 would suffice.
German has a huge resource but also in voxceleb has reasonable coverage as French seems to lack.
https://www.cl.uni-heidelberg.de/statnlpgroup/librivoxdeen/
Likely though you don't need to do by individual language but language group https://en.wikipedia.org/wiki/Language_family
Say for https://en.wikipedia.org/wiki/Indo-European_languages
Likely you could do a single model for France, Spain, Portugal, Italian & Romainian.
Another for English, German, Austrian, Swiss, Norway, Sweden, Iceland & Slavbard.
And so on but currently its hard to tell and guess the metadata must be somewhere and my hunch is right that it is Voxceleb 1&2 it could even be detrimental mixing these language families and definately so at such disportionate levels.
Likely Synthesian could leave his GPU's churning away on some fairly easy collatable datsets but grouped into better and specific language families.
Could be interesting grabbing and collating English, German, Austrian, Swiss, Norway, Sweden, Iceland & Slavbard into a single dataset and training an embedding model as see if for English it works better than current and turn out smaller.
Likely France, Spain, Portugal, Italian & Romainian model for those languages would work much better due to the low numbers in the current vs a large quantity of other language families.
Unlike ASR likely you could train Language families, actually could be the same with Whisper also as its 'Supposedly' multilingual but the smaller models become less so as a few people have mentioned the Tiny model for German is pretty stinky. Likely you could finetune Whisper by Language-Family also, without need for individual lang models.
https://www.speechly.com/blog/analyzing-open-ais-whisper-asr-models-word-error-rates-across-languages

If you where to do so it might be wise to start thinking of what pre-processing algs you use as likely say if beamforming 2 channel using a filter such as DTLN then grap your dataset and run through RIR GPU to create the 2Channel source then run through the beamforming and filter to create a resultant dataset to train and likely in use with 2channel beamforming and DTLN and that Language family the model will be even more accurate.
You can not really stick reverberation onto wavs in a single channel because that is the problem of reverberation as the sum you hear when mixed often as spectra looks like something else. You can but effect is limited and its much better to reduce incoming reverberation than try embed reverberation into a single channel because that is what reverberation does its a load of differing time signals mixing dependent on room shape, mics and furnishing and even at minimal levels you are making spectra less unique and fuzzy, by compensating in the datset than trying to remove preprocessing.
